File: cpp_code/odes/lib/set_boundary.cpp
#include "tensors.h"

// periodic boundary conditions here:
void set_boundary( dTensorBC2& q )
{

    const int mx   = q.get_size(1);
    const int meqn = q.get_size(2);
    const int mbc  = q.get_mbc();

    // TODO!  INCORPORATE BOUNDARY CONDITIONS OTHER THAN PERIODIC!
    int mbc_type [] = {0, 0};

    if( mbc_type[0] == 0 )
    { 
        for( int i=1; i <= mbc; i++ )
        for( int m=1; m <= meqn; m++ )
        {
            // right boundary cells:
            q.set( mx+i, m, q.get( i, m ) );

            // left boundary cells:
            q.set( 1-i,   m, q.get( (mx+1)-i, m ) );

        }
        return;
    }

    // dirichlet on left hand side:
    if( mbc_type[0] == 1 )
    {

        // left hand boundary values
        dTensor1 A(meqn);
        for( int m=1; m <= meqn; m++ )
        { A.set(m,0.0); }

        for( int m=1; m <= meqn; m++ )
        {


            // left hand boundary values (TODO - CALL A FUNCTION TO SET THIS?)
            if( mbc == 1 )
            {
                q.set(0, m, 2.0*A.get(m) - q.get(1,m) );
            } 
            else if( mbc == 2 )
            {
                q.set(0, m, 
                    4.0*A.get(m) - 13.0/3.0*q.get(1,m) +
                    5.0/3.0*q.get(2,m)-1.0/3.0*q.get(3,m) );
                q.set(-1, m, 
                    -12.0*A.get(m) + 7.0*( q.get(1,m) + q.get(0,m) )
                    - q.get(2,m) );
            }

        }
    }

    // dirichlet on right hand side:
    if( mbc_type[1] == 1 )
    {

        // right hand boundary values
        dTensor1 B(meqn);
        for( int m=1; m <= meqn; m++ )
        { B.set(m,0.0); } 

        for( int m=1; m <= meqn; m++ )
        {


            if( mbc == 1 )
            {
                q.set(mx+1, m, 2.0*B.get(m) - q.get(mx,1) );
            } 
            else if( mbc == 2 )
            {
                q.set(mx+1, m, 
                    4.0*B.get(m) + 1.0/3.0*( -13.0*q.get(mx,m) +
                    5.0*q.get(mx-1,m) - q.get(mx-2,m) ) );

                q.set(mx+2, m, 
                    4.0*B.get(m) + 1.0/3.0*( -13.0*q.get(mx+1,m) +
                    5.0*q.get(mx,m) - q.get(mx-1,m) ) );

            }

        }
    }

    // outflow on right hand side
    if( mbc_type[1] == 2 )
    {

        for( int m=1; m <= meqn; m++ )
        {

            if( mbc == 1 )
            {
                q.set(mx+1, m, 2.0*q.get(mx,1) - q.get(mx-1,m) );
            } 
            else if( mbc == 2 )
            {

                double tmp1 = -q.get(mx-3,m)+4.0*q.get(mx-2,m)
                              -6.0*q.get(mx-1,m)+4.0*q.get(mx,m);

                double tmp2 = -q.get(mx-2,m)+4.0*q.get(mx-1,m)
                              -6.0*q.get(mx,m)+4.0*q.get(1,m);

                q.set( mx+1, m, tmp1 );
                q.set( mx+2, m, tmp2 );

            }

        }
    }


}

File: matlab/pdes/lib/set_boundary.m
function [ql, qr] = set_boundary( q, t )

    global params

    mbc    = params.mbc;
    meqn   = params.meqn;
    sorder = params.sorder;
    mx     = params.mx;

    ql = zeros(mbc,meqn);
    qr = zeros(mbc,meqn);

    % periodic means periodic!
    if( strcmp( params.mbc_type_left , 'periodic' ) )

        % Trivial copy
        for i=1:mbc

            ql(i, :)  = q(mx-mbc+i, :);
            qr(i, :)  = q(i, : );

        end

        return;

    end

    % boundary condition for hte left hand side
    if( strcmp( params.mbc_type_left, 'dirichlet' ) )

        A = 0.0;

        if( mbc == 1 )

            % inflow
            ql(1,:) = 2*A-q(1,:);

            % outflow
            qr(1,:) = 2*q(mx,:) - q(mx-1,:);

        elseif( mbc == 2 )

            % inflow
            ql(2,:) =  4*A - 13/3*q(1,:) +  5/3*q(2,:)-1/3*q(3,:);

            ql(1,:) = 16*A - 70/3*q(1,:) + 32/3*q(2,:)-7/3*q(3,:);
            %           ql(1,:) =-12*A + 7*( ql(2,:)+q(1,:) ) - q(2,:);

        end

    elseif( strcmp( params.mbc_type_left, 'outflow' ) )

        disp('you need to be written!');

    else
        disp('unrecognized mbc for left hand side');
    end

    % Boundary condition for the right hand side
    if( strcmp( params.mbc_type_right, 'dirichlet' ) )

        B = 0.0;

        % diriclet!
        qr(1,:) = 4*B +1/3*(-13*q(mx,:) + 5*q(mx-1,:) -   q(mx-2,:) );
        qr(2,:) = 16*B+1/3*(-70*q(mx,:) +32*q(mx-1,:) - 7*q(mx-2,:) );


    elseif( strcmp( params.mbc_type_right, 'outflow' ) )

        % outflow
        % conditions from my maple script
        qr(1,:) = -q(mx-3,:)+4*q(mx-2,:)-6*q(mx-1,:)+4*q(mx,:);
        qr(2,:) = -q(mx-2,:)+4*q(mx-1,:)-6*q(mx,:)+4*qr(1,:);

    else
        disp('unrecognized mbc for left hand side');
    end

end
